java/ql/src/semmle/code/java/dataflow/FlowSources.qll

```diff
@@ -270,3 +319,49 @@ class AndroidIntentInput extends DataFlow::Node {
     )
   }
 }
+
+/** Method access to external inputs of `android.content.Intent` object. */
+class IntentGetExtraMethodAccess extends MethodAccess {
+  IntentGetExtraMethodAccess() {
+    exists(AndroidComponent ac |
+      this.getEnclosingCallable().getDeclaringType() = ac and ac.isExported()
+    ) and
+    (
+      this.getMethod().getName().regexpMatch("get\\w+Extra") and
+      this.getMethod().getDeclaringType() instanceof TypeIntent
+      or
+      this.getMethod().getName().regexpMatch("get\\w+") and
+      this.getQualifier().(MethodAccess).getMethod().hasName("getExtras") and
+      this.getQualifier().(MethodAccess).getMethod().getDeclaringType() instanceof TypeIntent
+    )
+  }
+}
+
+/** Android intent extra source. */
+private class AndroidIntentExtraSource extends RemoteFlowSource {
```

Comment:
At the moment this is quite brittle: it particularly wants the flow from the Intent to `getExtras()` to be by direct use as a qualifier, for example. This is manually modelling one-step dataflow.

It would be better to make `AndroidIntentInput` a source, then implement `AdditionalTaintStep` to note that if the intent is tainted, so is the output of `getStringExtra`, and also `getExtras`, and in turn if the Extras datastructure is tainted so are its accessors.

That allows arbitrary dataflow edges to come between each step -- the intent can be passed into a function, that retrieves the Extras, then passes *that* into a different function, which calls getString... and so on.

The Java library has so far always done this per-query, like this (https://github.com/github/codeql/blob/main/java/ql/src/semmle/code/java/security/LdapInjection.qll#L390), but there's no reason we shouldn't model query-independent flow steps like this by directly extending AdditionalTaintStep (https://github.com/github/codeql/blob/75357727c49a968c15050390f8aee695b0b6190a/java/ql/src/semmle/code/java/dataflow/FlowSteps.qll#L26)

Reply:
Thanks @smowton for the advice. I've revamped the query to implement <code>AdditionalTaintStep</code> to allow arbitrary dataflow edges to come between each step.

Then a query like <code>UnsafeAndroidAccess.ql</code> that I submitted can be changed to the following to use <code>RemoteFlowSource</code> with <code>isAdditionalTaintStep</code>:
<code>
class FetchUntrustedResourceConfiguration extends TaintTracking::Configuration {
...
  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
...
  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
    any(AndroidExtraSourceAdditionalTaintStep a).step(pred, succ)
  }
</code>

Please let me know if there are still some issues.

Cheers,
@luchua-bc